### ISSUE TITLE
late final local variable re-initialized multiple times

### ISSUE THREAD

**Issue by yjbanov** [dart-lang/sdk#42422]
Repro:

```dart
class _EllipseSegmentResult {
  _EllipseSegmentResult() {
    print('>>> _EllipseSegmentResult constructor');
  }
  double endPointX = 0;
  double endPointY = 0;
  double distance = 0;
}

void main() {
  late final _EllipseSegmentResult result = _EllipseSegmentResult();
  for (int i in [1, 2, 3]) {
    print(result.endPointX);
  }
}
```

Expected output:

```
>>> _EllipseSegmentResult constructor
0
0
0
```

Actual output:

```
>>> _EllipseSegmentResult constructor
0
>>> _EllipseSegmentResult constructor
0
>>> _EllipseSegmentResult constructor
0
```

I compiled with `dart2js` without minification and without inlining.

**Comment by rakudrama** [comment]
When run with dart2js_developer there are broken invariants:
```
Unhandled exception:
'package:compiler/src/ssa/locals_handler.dart': Failed assertion: line 423 pos 12: '!isStoredInClosureField(local)': Local local(main1#result) is stored in a closure field.
#0      _AssertionError._doThrowNew (dart:core-patch/errors_patch.dart:46:39)
#1      _AssertionError._throwNew (dart:core-patch/errors_patch.dart:36:5)
#2      LocalsHandler.updateLocal (package:compiler/src/ssa/locals_handler.dart:423:12)
#3      KernelSsaGraphBuilder._visitLocalSetter (package:compiler/src/ssa/builder_kernel.dart:3510:19)
#4      KernelSsaGraphBuilder.visitVariableSet (package:compiler/src/ssa/builder_kernel.dart:3472:5)
#5      VariableSet.accept (package:kernel/ast.dart:3056:44)
#6      KernelSsaGraphBuilder.visitConditionalExpression.<anonymous closure> (package:compiler/src/ssa/builder_kernel.dart:2958:25)
#7      SsaBranchBuilder.buildBranch (package:compiler/src/ssa/ssa_branch_builder.dart:93:16)
```
There is almost no difference between the Kernel code for these two:
```dart
void main1() {
  late final _EllipseSegmentResult result = _EllipseSegmentResult();
  for (int i in [1, 2, 3]) {
    print(result.endPointX);
  }
}

void main2() {
  _EllipseSegmentResult? _result;
  _EllipseSegmentResult result() {
    var r = _result;
    return r == null ? (_result = _EllipseSegmentResult()) : r;
  }
  for (int i in [1, 2, 3]) {
    print(result().endPointX);
  }
}
```
`main2` runs as expected. There is something wrong with scope analysis and closure conversion.

**Comment by rakudrama** [comment]
Fix on the way: https://dart-review.googlesource.com/c/sdk/+/151867

**Comment by sigmundch** [comment]
seems like the fix landed in 8e2c51e8fd0c9203ae29b7f916937ce8673223ec, so we can now close this.
